tests/cacl/test_cacl_application.py

```diff
@@ -59,7 +63,128 @@ def docker_network(duthost):
+
+    Args:
+        duthosts: All DUTs belong to the testbed.
+        rand_one_dut_hostname: hostname of a random chosen dut to run test.
+        collect_ignored_rules: ignored iptable/ip6table rules.
+
+    Returns:
+        None
+    """
+    duthost = duthosts[rand_one_dut_hostname]
+
+    yield
+
+    logger.info("delete tmp file and recover ACL configuration")
+    # delete the tmp file
+    duthost.file(path=SCALE_ACL_FILE, state='absent')
+    # recover ACL configuration
+    duthost.command("acl-loader delete SNMP_ACL")
+    duthost.command("acl-loader delete NTP_ACL")
+    duthost.command("acl-loader delete SSH_ONLY")
+
+    logger.info('Waiting all rules to be cleaned')
+    wait_until(60, 2, 2, is_acl_rule_empty, duthost)
 
+    # acl-loader delete operation only deletes acl rules, it still has to wait some time for synchronizing iptables rules
+    if duthost.is_multi_asic:
+        time.sleep(60)
+    else:
+        time.sleep(10)
+
+    # remove the SSH ACCEPT rule
+    iptables_rule_list = duthost.command("iptables -L -n --line-number")["stdout_lines"]
+    if "3   ACCEPT     tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:22" in iptables_rule_list:
+        duthost.command("iptables -D INPUT 3")
+
+    # The following code to check if clean CACL rules are cleaned clearly
+    ignored_rules_v4 = duthost.command("iptables -S")["stdout_lines"]
+    ignored_rules_v6 = duthost.command("ip6tables -S")["stdout_lines"]
+
+    try:
+        uncleaned_iptables_rules = set(ignored_rules_v4) - set(collect_ignored_rules["v4"])
+        pytest_assert(len(uncleaned_iptables_rules) == 0, "Some iptables rules are not cleaned: {}".format(repr(uncleaned_iptables_rules)))
+        uncleaned_ip6tables_rules = set(ignored_rules_v6) - set(collect_ignored_rules["v6"])
+        pytest_assert(len(uncleaned_ip6tables_rules) == 0, "Some ip6tables rules are not cleaned: {}".format(repr(uncleaned_ip6tables_rules)))
+    except:
+        logger.info("Some iptalbes/ip6tables rules are not cleaned clearly. Config reload to recover testbed and make sure not impact the following tests.")
+        config_reload(duthost, wait=60)
+        raise
+
+
+def is_acl_rule_empty(duthost):
+    """
+    Check the output of "show acl rule", return True if rules are cleaned.
+
+    Args:
+        duthosts: All DUTs belong to the testbed.
+
+    Returns:
+        boolean: True of False
+    """
+    stdout_lines = duthost.command("show acl rule")["stdout_lines"]
+
+    stdout_lines = stdout_lines[2:]
+    if len(stdout_lines) != 0:
+        return False
+
+    return True
+
+def check_iptable_rules(duthost):
+    """
+    It's used to keep ssh session not timeout, otherwise reconnection will be failed due to default CACL DENY rule.
+
+    Args:
+        duthosts: All DUTs belong to the testbed.
+
+    Returns:
+        boolean: False
+    """
+    duthost.command("iptables -S")
+    duthost.command("ip6tables -S")
+
+    return False
```

Comment:
What's the purpose of `check_iptable_rules`? It always return False?

Reply:
As I mentioned in summary, "acl-loader update full **.json" command will refresh iptables, we have to add the ACCEPT SSH iptables rule after acl-loader command. But on multi-asic testbed, it always costs minutes to sync iptables rules after updating cacl rules, at first, I added time.sleep to make the process to wait, but the default ansible SSH timeout is 30s, after sleeping for 3 mins, the process will then try to add the ACCEPT SSH rule, but SSH connection is disconnected now, duthost.command will try to reconnect to DUT, at this time, next duthost.command will trigger ssh login, it will be rejected by default CACL DENY rule, test will fail. We have wait_until to solve this problem, I add wail_until to call check_iptable_rules every 10s to keep session alive, in check_iptable_rules just uses duthost.command to active ssh connection. In this way, we can active ssh connection and wait how long as we want.